src/hotspot/share/gc/g1/g1FullGCPrepareTask.cpp

```diff
@@ -61,13 +62,25 @@ bool G1FullGCPrepareTask::G1CalculatePointersClosure::do_heap_region(HeapRegion*
     }
   } else {
     assert(!hr->is_humongous(), "moving humongous objects not supported.");
-    prepare_for_compaction(hr);
+    size_t live_words = _collector->hr_live_words(hr->hrm_index());
+    size_t live_words_threshold = _collector->scope()->hr_live_words_threshold();
+
+    if(live_words <= live_words_threshold) {
+      prepare_for_compaction(hr);
+    } else {
+      assert(MarkSweepDeadRatio > 0,
+                "it should not trigger skipping compaction, when MarkSweepDeadRatio == 0");
+      // Force the high live ration region pinned,
+      // as we need skip these regions in the later compact step.
+      force_pinned = true;
+      log_debug(gc, phases)("Phase 2: skip compaction region index: %u, live words: " SIZE_FORMAT,
+                            hr->hrm_index(), live_words);
+    }
```

Comment:
Doing the above, all this can go back to just being:
`prepare_for_compaction(hr);`

And you can add the logging in the pinned section above, something like:
```
else if(hr->is_pinned_for_gc()) {
  assert(MarkSweepDeadRatio > 0,
         "it should not trigger skipping compaction, when MarkSweepDeadRatio == 0");
  log_debug(gc, phases)("Phase 2: skip compaction region index: %u, live words: " SIZE_FORMAT,
                        hr->hrm_index(), live_words);
}